build.assets/Dockerfile-multiarch

```diff
@@ -50,25 +26,213 @@ RUN groupadd ci --gid=$GID -o && \
 
 RUN install --directory --mode=0700 --owner=ci --group=ci /var/lib/teleport
 
-ENV LANGUAGE=en_US.UTF-8 \
-    LANG=en_US.UTF-8 \
-    LC_ALL=en_US.UTF-8 \
-    LC_CTYPE=en_US.UTF-8
-
-RUN yum update -y $YUM_UPDATE_FLAGS && \
-    yum install -y     \
-        git \
-        gcc \
-        gcc-c++ \
-        make \
-        pam-devel \
-        perl-IPC-Cmd \
-        tree \
-        which \
-        zip \
-        zlib-static && \
-    yum clean all && \
-    localedef -c -i en_US -f UTF-8 en_US.UTF-8
+## LIBPCSCLITE ################################################################
+# 
+FROM gcc AS libpcsclite
+ARG LIBPCSCLITE_VERSION
+
+ENV CFLAGS="-std=gnu99"
+
+# Install libpcsclite - compile with a newer GCC. The one installed by default is not able to compile it.
+RUN git clone --depth=1 https://github.com/gravitational/PCSC.git -b ${LIBPCSCLITE_VERSION} && \
+    cd PCSC && \
+    ./bootstrap && \
+    ./configure --enable-static --with-pic --disable-libsystemd && \
+    make install && \
+    rm -rf PCSC
+
+## LIBFIDO2 ###################################################################
+#
+
+# Build libfido2 separately for isolation, speed and flexibility.
+FROM gcc AS libfido2
+
+# Install libudev-zero.
+# libudev-zero replaces systemd's libudev.
+RUN git clone --depth=1 https://github.com/illiliti/libudev-zero.git -b 1.0.1 && \
+    cd libudev-zero && \
+    [ "$(git rev-parse HEAD)" = '4154cf252c17297f98a8ca33693ead003b4509da' ] && \
+    make install-static LIBDIR='$(PREFIX)/lib64'
+
+# Instal openssl.
+# Pulled from source because repository versions are too old.
+# install_sw install only binaries, skips docs.
+RUN git clone --depth=1 https://github.com/openssl/openssl.git -b OpenSSL_1_1_1t && \
+    cd openssl && \
+    [ "$(git rev-parse HEAD)" = '830bf8e1e4749ad65c51b6a1d0d769ae689404ba' ] && \
+    ./config --release --libdir=/usr/local/lib64 && \
+    make && \
+    make install_sw
+
+# Install libcbor.
+RUN git clone --depth=1 https://github.com/PJK/libcbor.git -b v0.10.2 && \
+    cd libcbor && \
+    [ "$(git rev-parse HEAD)" = 'efa6c0886bae46bdaef9b679f61f4b9d8bc296ae' ] && \
+    cmake3 \
+        -DCMAKE_CXX_FLAGS=-lpthread \
+        -DCMAKE_BUILD_TYPE=Release \
+        -DCMAKE_POSITION_INDEPENDENT_CODE=ON \
+        -DWITH_EXAMPLES=OFF . && \
+    make && \
+    make install
+
+# Install libfido2.
+# Depends on libcbor, openssl, zlib-devel and libudev.
+# Linked so `make build/tsh` finds the library where it expects it.
+RUN git clone --depth=1 https://github.com/Yubico/libfido2.git -b 1.12.0 && \
+    cd libfido2 && \
+    [ "$(git rev-parse HEAD)" = '659a02679f99fd34a44e06e35dce90794f6ecc86' ] && \
+      LDFLAGS="-Wl,--copy-dt-needed-entries" cmake3 \
```

Comment:
Why `--copy-dt-needed-entries` is needed here? I don't see that is the case on master.

CC @codingllama

Reply:
Beats me too.

Reply:
#20 22.74 /usr/bin/ld: /usr/local/lib/gcc/x86_64-unknown-linux-gnu/10.4.0/../../../../lib64/libudev.a(udev_enumerate.o): undefined reference to symbol 'pthread_create@@GLIBC_2.2.5'
#20 22.74 //usr/lib64/libpthread.so.0: error adding symbols: DSO missing from command line
(ARM64)

> With --copy-dt-needed-entries dynamic libraries mentioned on the command line will be recursively searched, following their DT_NEEDED tags to other libraries, in order to resolve symbols required by the output binary. With the default setting however the searching of dynamic libraries that follow it will stop with the dynamic library itself. No DT_NEEDED links will be traversed to resolve symbols.

Reply:
Sounds like a wrong linking order. Can you try to append `-lpthread` instead of "recursive search". Sounds like only this one library is missing and this recursive search can do some damage in rare cases.

Reply:
Okay, will try.

Reply:
Yes, you were correct. It works with -lpthread, my fault here.